Don't allow removing the policy owner from workspace members

pecanoro opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
Disable the checkbox and shows a tooltip if the admin tries to remove the owner or themselves from the workspace.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/179467

### Tests / QA

1. In an account that has an existing workspace (or you can create a new one), click on People.
2. Add a couple of extras users to the workspace.
3. Make sure you can select all users to be removed except for the owner of the workspace and yourself.
4. If you hover on web or desktop on the owner or yourself, you should get the tooltip shown in the screenshots.
5. If you test this on a mobile device, a growl should show up instead (as shown in the screenshots)
6. Play with the `Select All` button and make sure it still works, it should select all users except for the above.
7. Test that removing the selected users work as expected and they get removed.

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

### Screenshots
<!-- Add screenshots for all platforms tested. Pull requests won't be merged unless the screenshots show the app was tested on all platforms.-->

#### Web
![New Expensify 2021-09-30 17-30-13](https://user-images.githubusercontent.com/6474442/135499859-c6b12e33-49f5-4cba-88c8-a8bb627dc004.png)

#### Mobile Web
![iPhone 12 2021-10-04 15-13-02](https://user-images.githubusercontent.com/6474442/135877734-28b796c9-159e-4489-9aac-46e695a5f61e.png)

#### Desktop
![New Expensify 2021-09-30 17-30-13](https://user-images.githubusercontent.com/6474442/135499878-1f05a896-fdcc-4b5e-a9f8-878b03b8b129.png)

#### iOS
![iPhone 12 2021-10-04 12-16-14](https://user-images.githubusercontent.com/6474442/135825603-ce9f99bf-0cbe-4b63-84b0-81ba2e80c896.png)

#### Android
![Android Emulator - Pixel_5_API_30:5554 2021-10-04 11-36-42](https://user-images.githubusercontent.com/6474442/135820799-84f1f2ca-b169-406e-a34d-4447e94720f9.png)

pecanoro committed:
Remmove checkbox for policy owner

pecanoro committed:
Disable checkbox  instead

pecanoro committed:
Disable whole row

pecanoro committed:
Show tooltip on checkbox

pecanoro committed:
Fix select all

pecanoro committed:
Merge branch 'main' of github.com:Expensify/App into rocio-RemoveMember

pecanoro committed:
Merge branch 'main' of github.com:Expensify/App into rocio-RemoveMember

pecanoro committed:
Play with modal instead

pecanoro committed:
Show growl on mobile

pecanoro committed:
Make it work for mobile browsers

tgolen commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -169,17 +183,32 @@ class WorkspacePeoplePage extends React.Component {
     }) {
+        const initialDimensions = Dimensions.get('window');
+        const isSmallScreenWidth = initialDimensions.width <= variables.mobileResponsiveWidthBreakpoint;
+        const isMobileDevice = getPlatform() === CONST.PLATFORM.ANDROID || getPlatform() === CONST.PLATFORM.IOS || isSmallScreenWidth;
```
`getPlatform()` is a restricted method that should never be used inside a component. The proper way to handle this is to create platform specific files (eg. `.ios.js`, `.android.js`, etc.). 

You could perhaps do this with a new component called `CheckBoxWithTooltip` and moving this logic inside that new component.

I also think that `renderItem()` should be moved to its own file to help with this.

Lastly, there is a higher order component called `withWindowDimensions` that could be useful.

pecanoro replied:
Ok!! I will do that!!

Julesssss commented on src/pages/workspace/WorkspacePeoplePage.js:
```diff
@@ -169,17 +183,32 @@ class WorkspacePeoplePage extends React.Component {
     }) {
+        const initialDimensions = Dimensions.get('window');
+        const isSmallScreenWidth = initialDimensions.width <= variables.mobileResponsiveWidthBreakpoint;
+        const isMobileDevice = getPlatform() === CONST.PLATFORM.ANDROID || getPlatform() === CONST.PLATFORM.IOS || isSmallScreenWidth;
```
Reminder for us to test this on iPad, as I'm not sure whether `isSmallScreenWidth` will include large iPads.

pecanoro committed:
Create new component

pecanoro committed:
Move code to components

pecanoro committed:
Linter errors

pecanoro committed:
Fix errorss

pecanoro commented:
@tgolen I moved it to a component like you mentioned, but I encountered the following roadblocks and maybe it's just that I am not super strong in React Native:

1. Before I moved the checkbox and the tooltip to a component, I was able to use the TouchableOpacity to trigger the tooltips and growls wherever the user clicked in the row, but I can't do that anymore and I am not sure how to do it since it's a higher level than the new component.
2. I have no idea how to show the growl in mobile web browsers, I don't there is a specific file like native.js for that. If I add it to index.js, I would end up just doing what I was doing before, checking for the dimensions and I would be replicating the logic in native.js but in that file, not great for DRY.

Any advice on how I could solve these problems?

tgolen commented:
OK, let's see if I can help here. For the first problem, I would start with changing the onpress handler to be more like this:

```diff
- onPress={canBeRemoved ? () => this.toggleUser(item.login) : () => {}}
+ onPress={() => this.toggleUser(item.login)}
```

Then update toggleuser to be more like this:
```
    toggleUser(login) {
        const canBeRemoved = this.props.policy.owner !== item.login && this.props.session.email !== item.login;
        if (!canBeRemoved) {
            this.setState({showTooltipForLogin: login});
            return;
        }
        // Add or remove the user if the checkbox is enabled and is clickable.
        if (_.contains(this.state.selectedEmployees, login)) {
            this.removeUser(login);
        } else {
            this.addUser(login);
        }
    }
```

Then finally update the checkbox component:
```diff
- toggleTooltip={!canBeRemoved}
+ toggleTooltip={this.state.showTooltipForLogin === login}
```

Would that work?

For your second issue, I don't think I am understanding the problem. There shouldn't be anything specific to mobile web for showing a growl and you can just use `Growl.show()`. Are you saying it's broken for mobile web but works on all other platforms? If so, is it that way in `main` too? Maybe it's not something specific to your PR.

pecanoro committed:
Show for whole row

pecanoro committed:
Show growl on tap

pecanoro commented:
> There shouldn't be anything specific to mobile web for showing a growl and you can just use Growl.show(). Are you saying it's broken for mobile web but works on all other platforms? If so, is it that way in main too? Maybe it's not something specific to your PR.

@tgolen Let me explain again, so mobile browsers on phones use index.js and not index.native.js, so it will try to show a tooltip as well, but it won't work. The only way to do this (I think) would be to check for the dimensions. But then I feel I would be repeating most of the code from index.js to index.native.js

pecanoro committed:
Check for dimensions

pecanoro committed:
Import native component

pecanoro committed:
Remove docs

pecanoro committed:
Merge branch 'main' of github.com:Expensify/App into rocio-RemoveMember

pecanoro committed:
Refactor component

pecanoro committed:
Merge branch 'main' into rocio-RemoveMember

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipForMobileWebAndNative.js:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Show warning modal on mobile devices since tooltips are not supported when checkbox is disabled.
+     */
+    onPress() {
```
This method should be named for what it does (not what it handles). The only name I could come up with is `showGrowlAndTriggerOnPress`

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipForMobileWebAndNative.js:
```diff
@@ -0,0 +1,49 @@
+        if (this.props.toggleTooltip) {
+            Growl.show(this.props.text, CONST.GROWL.WARNING, 3000);
+        }
+        this.props.onPress();
```
Should this still be triggered when the growl is shown?

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipForMobileWebAndNative.js:
```diff
@@ -0,0 +1,49 @@
+     */
+    onPress() {
+        if (this.props.toggleTooltip) {
+            Growl.show(this.props.text, CONST.GROWL.WARNING, 3000);
```
I think in the context of this PR, the WARNING growl makes sense, but since this is a generic component, I don't know that it is good to have this default to WARNING. It would mean that any checkbox with a tooltip would always display the growl as a warning. Is that what we want?

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipForMobileWebAndNative.js:
```diff
@@ -0,0 +1,49 @@
+CheckboxWithTooltipForMobileWebAndNative.propTypes = propTypes;
+CheckboxWithTooltipForMobileWebAndNative.defaultProps = defaultProps;
+
+export default compose(
```
Compose shouldn't be used when there is only one method (same goes for the other components in this PR).

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipForMobileWebAndNative.js:
```diff
@@ -0,0 +1,49 @@
+     * Show warning modal on mobile devices since tooltips are not supported when checkbox is disabled.
+     */
+    onPress() {
+        if (this.props.toggleTooltip) {
```
Should this also check that the checkbox is disabled?

tgolen commented on src/components/CheckboxWithTooltip/CheckboxWithTooltipPropTypes.js:
```diff
@@ -0,0 +1,36 @@
+    disabled: PropTypes.bool,
+
+    /** Container styles */
+    style: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.object), PropTypes.object]),
```
There is a `stylePropTypes.js` that you can use for this instead.

tgolen reviewed (changes requested):
Those changes to the component look great!

pecanoro committed:
Address comments

pecanoro committed:
Style

pecanoro committed:
Fix bug

pecanoro commented:
I fixed the bug!! I had a revelation this morning to use hoverable to fix the tooltip so it would work not only on press! Mobiles devices don't have hoverables, so win-win.

pecanoro committed:
Only toggle on disabled

Julesssss commented on src/pages/workspace/WorkspaceMembersPage.js:
```diff
@@ -169,41 +196,46 @@ class WorkspaceMembersPage extends React.Component {
+                <TouchableOpacity
+                    style={[styles.peopleRow, !canBeRemoved && styles.cursorDisabled]}
+                    onPress={() => this.toggleUser(item.login)}
+                    activeOpacity={0.7}
```
Nevermind, I see this was already added.

Julesssss reviewed (changes requested):
It's working well on all platforms, but I noticed that we're showing the Growl on hover when the website window is 'small'. I think this is not expected, right? In this case, shouldn't we just display the growl on touch? This doesn't seem to occur on Desktop.

https://user-images.githubusercontent.com/10736861/137150739-0273a20b-557d-48a3-9275-27180a7484ae.mov

pecanoro committed:
Fix small screen bug

pecanoro commented:
@Julesssss had an idea to fix the weird bug and it worked!! So updated again with the small changes

Julesssss reviewed (approved):
👍

pecanoro commented:
I double-checked on Slack, n6 polish issues are ok to merge so self-merging since everyone approved.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/1374284625) to production by @roryabraham in version: 1.1.8-9 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
